### PR TITLE
fixed validity and renewal days propagation to user operator

### PR DESCRIPTION
**Pull request by stanlyDoge**
### Type of change
- Bugfix

### Description
User operator was ignoring changes in validity and renewal days as described in https://github.com/strimzi/strimzi-kafka-operator/issues/1240

### Checklist
- [ ] Update/write design documentation in `./design`
- [x] Write tests
- [x] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java [hunk 1]
```diff
@@ -42,12 +45,16 @@
...
+    public static final String ENV_VAR_CLIENTS_CA_VALIDITY = "STRIMZI_CA_VALIDITY";
+    public static final String ENV_VAR_CLIENTS_CA_RENEWAL = "STRIMZI_CA_RENEWAL";
 
     private String zookeeperConnect;
     private String watchedNamespace;
     private String resourceLabels;
     private long reconciliationIntervalMs;
     private long zookeeperSessionTimeoutMs;
+    private long clientsCaValidityDays;
+    private long clientsCaRenewalDays;
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java]: These things should not be reconfigured in `Kafka.spec.entityOperator.userOperator`, but rather be taken from `Kafka.spec.clientsCa` implicitly.

Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java]: I am afraid I do not understand :( It takes values from ```result.setClientsCaValidityDays(kafkaAssembly.getSpec().getClientsCa().getValidityDays());
                    result.setClientsCaRenewalDays(kafkaAssembly.getSpec().getClientsCa().getRenewalDays());```
What do you mean by reconfiguring?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java [hunk 2]
```diff
@@ -34,6 +34,9 @@
     protected static final int HEALTHCHECK_PORT = 8081;
     protected static final String HEALTHCHECK_PORT_NAME = "healthcheck";
 
+    public final static int DEFAULT_CERTS_VALIDITY_DAYS = 365;
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java]: Shoudn't the defaults be set in the `api`? Wasn't that how we were doing it lately?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java [hunk 3]
```diff
@@ -167,6 +192,10 @@ public static EntityUserOperator fromCrd(Kafka kafkaAssembly) {
                 result.setLogging(userOperatorSpec.getLogging());
                 result.setGcLoggingEnabled(userOperatorSpec.getJvmOptions() == null ? true : userOperatorSpec.getJvmOptions().isGcLoggingEnabled());
                 result.setResources(userOperatorSpec.getResources());
+                if (kafkaAssembly.getSpec().getClientsCa() != null) {
+                    result.setClientsCaValidityDays(kafkaAssembly.getSpec().getClientsCa().getValidityDays());
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java]: What if the the `clientsCa` exists, but the `validityDays and `renewalDays` are not set? You would just overwrite the defaults here with null, or?

File: user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java [hunk 4]
```diff
@@ -133,13 +134,23 @@ public void maybeGenerateCertificates(CertManager certManager,
         } else if (clientsCaKeySecret == null) {
             throw new NoCertificateSecretException("The Clients CA Key Secret is missing");
         } else {
+            String valStr = System.getenv(ENV_VAR_CLIENTS_CA_VALIDITY);
```
Comment (scholzj) [inline user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java]: You should not be getting the env vars here but rather in `UserOperatorConfig` where we take and process other env vars as well. That will allow us to have them all in one place.

File: user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java [hunk 5]
```diff
@@ -134,15 +135,9 @@ public void maybeGenerateCertificates(CertManager certManager,
...
-            String valStr = System.getenv(ENV_VAR_CLIENTS_CA_VALIDITY);
-            String renStr = System.getenv(ENV_VAR_CLIENTS_CA_RENEWAL);
-
-            int validity = 0;
-            int renewal = 0;
-            if (valStr != null)
-                validity = Integer.parseInt(valStr);
-            if (renStr != null)
-                renewal = Integer.parseInt(renStr);
+            UserOperatorConfig config = UserOperatorConfig.fromMap(System.getenv());
```
Comment (scholzj) [inline user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java]: I'm not sure how expensive operation this is. But having it here means it can be possibly called many many times. Can't we pass it somehow from the `UserOperatorConfig` instance created at start?

@tombentley What do you think about this?

Comment (tombentley) [reply user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java]: @scholzj is right, these should be parameters to `fromCrd()` and passed from there to `maybeGenerateCertificates()`. Also `maybeGenerateCertificates()` (and `maybeGeneratePassword()`) should be `private`.

Comment (scholzj) [reply user-operator/src/main/java/io/strimzi/operator/user/model/KafkaUserModel.java]: TBH, what I meant as a solution was to just pass the today variables through the objects. Not make them static methods somewhere else. You will still do the `System.getEnv()` again and again and again although it doesn't really change. Although [SO](https://stackoverflow.com/questions/42196291/the-performance-impact-of-use-system-getenv-in-java) seems to suggest it is cheap operation. I will leave it up to @tombentley to see if that is or is not an issue.

File: api/src/main/java/io/strimzi/api/kafka/model/EntityUserOperatorSpec.java [hunk 6]
```diff
@@ -39,6 +39,8 @@
     public static final int DEFAULT_ZOOKEEPER_PORT = 2181;
     public static final long DEFAULT_FULL_RECONCILIATION_INTERVAL_SECONDS = 120;
     public static final long DEFAULT_ZOOKEEPER_SESSION_TIMEOUT_SECONDS = 6;
+    public static final int DEFAULT_CERTS_VALIDITY_DAYS = 365;
+    public static final int DEFAULT_CERTS_RENEWAL_DAYS = 30;
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/EntityUserOperatorSpec.java]: We are starting to have these two values all around the operators / api.
The `ModelUtils` class for example is using the constant from the `AbstractModel` about the cert validity days while it has an hard coded "30" value for the cert renewal days.
Are the above constants in the `api` module because we want them to be available through the api? In that case, because the validity is the same is the `EntityUserOperatorSpec` the right place?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java [hunk 7]
```diff
@@ -198,6 +230,8 @@ public static EntityUserOperator fromCrd(Kafka kafkaAssembly) {
         varList.add(buildEnvVar(ENV_VAR_CLIENTS_CA_KEY_SECRET_NAME, KafkaCluster.clientsCaKeySecretName(cluster)));
         varList.add(buildEnvVar(ENV_VAR_CLIENTS_CA_CERT_SECRET_NAME, KafkaCluster.clientsCaCertSecretName(cluster)));
         varList.add(buildEnvVar(ENV_VAR_STRIMZI_GC_LOG_ENABLED, String.valueOf(gcLoggingEnabled)));
+        varList.add(buildEnvVar(ENV_VAR_CLIENTS_CA_VALIDITY, Long.toString(clientsCaValidityDays)));
+        varList.add(buildEnvVar(ENV_VAR_CLIENTS_CA_RENEWAL, Long.toString(clientsCaRenewalDays)));
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java]: Why are you using `Long` here but then `Integer` in the test. I guess `Integer` is ok for both.

File: api/src/main/java/io/strimzi/api/kafka/model/CertificateAuthority.java [hunk 8]
```diff
@@ -39,7 +39,7 @@
 
     @Description("The number of days generated certificates should be valid for. The default is 365.")
     @Minimum(1)
-    public int getValidityDays() {
+    public Integer getValidityDays() {
```
Comment (tombentley) [inline api/src/main/java/io/strimzi/api/kafka/model/CertificateAuthority.java]: Why was this necessary?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java [hunk 9]
```diff
@@ -82,7 +82,8 @@
 
     protected static final Logger log = LogManager.getLogger(AbstractModel.class.getName());
 
-    protected static final int CERTS_EXPIRATION_DAYS = 365;
+    protected static final int CERTS_VALIDITY_DAYS = 365;
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: If we're changing the name we should use `DEFAULT_CERTS_VALIDITY_DAYS`.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java [hunk 10]
```diff
@@ -82,7 +82,8 @@
...
+    protected static final int CERTS_RENEWAL_DAYS = 30;
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java]: `DEFAULT_CERTS_RENEWAL_DAYS`

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java [hunk 11]
```diff
@@ -167,6 +189,16 @@ public static EntityUserOperator fromCrd(Kafka kafkaAssembly) {
                 result.setLogging(userOperatorSpec.getLogging());
                 result.setGcLoggingEnabled(userOperatorSpec.getJvmOptions() == null ? true : userOperatorSpec.getJvmOptions().isGcLoggingEnabled());
                 result.setResources(userOperatorSpec.getResources());
+                if (kafkaAssembly.getSpec().getClientsCa() != null) {
+                    if (kafkaAssembly.getSpec().getClientsCa().getValidityDays() != null) {
+                        log.debug("Setting validity days to " + kafkaAssembly.getSpec().getClientsCa().getValidityDays());
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java]: I don't think this really adds anything. None of the other places where we set a value in this method log anything.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java [hunk 12]
```diff
@@ -167,6 +189,16 @@ public static EntityUserOperator fromCrd(Kafka kafkaAssembly) {
...
+                        result.setClientsCaValidityDays(kafkaAssembly.getSpec().getClientsCa().getValidityDays());
+                    }
+                    if (kafkaAssembly.getSpec().getClientsCa().getRenewalDays() != null) {
+                        log.debug("Setting renewal days to " + kafkaAssembly.getSpec().getClientsCa().getRenewalDays());
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/EntityUserOperator.java]: Again, I don't think we need the logging.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java [hunk 13]
```diff
@@ -54,7 +54,7 @@ public static Secret findSecretWithName(List<Secret> secrets, String sname) {
     }
 
     public static int getCertificateValidity(CertificateAuthority certificateAuthority) {
-        int validity = AbstractModel.CERTS_EXPIRATION_DAYS;
+        int validity = AbstractModel.CERTS_VALIDITY_DAYS;
         if (certificateAuthority != null
                 && certificateAuthority.getValidityDays() > 0) {
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java]: Since you change it from primitive to boxed, there is now the possibility of an NPE here (though perhaps is never happens in practice).

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java [hunk 14]
```diff
@@ -63,7 +63,7 @@ public static int getCertificateValidity(CertificateAuthority certificateAuthori
     }
 
     public static int getRenewalDays(CertificateAuthority certificateAuthority) {
-        return certificateAuthority != null ? certificateAuthority.getRenewalDays() : 30;
+        return certificateAuthority != null ? certificateAuthority.getRenewalDays() : AbstractModel.CERTS_RENEWAL_DAYS;
```
Comment (tombentley) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java]: Again, possible NPE.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/EntityUserOperatorTest.java [hunk 15]
```diff
@@ -163,4 +166,37 @@ public void testGetContainers() {
...
+                        .withEntityOperator(entityOperatorSpec)
+                        .endSpec()
+                        .build();
+        EntityUserOperator entityUserOperator2 = EntityUserOperator.fromCrd(defaultValues);
+
+        assertEquals(42, entityUserOperator.getClientsCaValidityDays());
+        assertEquals(69, entityUserOperator.getClientsCaRenewalDays());
+        assertEquals(AbstractModel.CERTS_VALIDITY_DAYS, entityUserOperator2.getClientsCaValidityDays());
+        assertEquals(AbstractModel.CERTS_RENEWAL_DAYS, entityUserOperator2.getClientsCaRenewalDays());
+    }
```
Comment (tombentley) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/EntityUserOperatorTest.java]: What I don't see it a test covering the end-to-end scenario: That the UO `Deployment` ends up with the relevent env vars set.

File: user-operator/src/main/java/io/strimzi/operator/user/UserOperatorConfig.java [hunk 16]
```diff
@@ -115,6 +123,25 @@ public static UserOperatorConfig fromMap(Map<String, String> map) {
...
+
+    private static int getIntProperty(String name) {
+        // default value is set in operator
+        Map<String, String> map = System.getenv();
+        int value = 0;
+        String intEnvVar = map.get(name);
+        if (intEnvVar != null) {
+            value = Integer.parseInt(intEnvVar);
+        }
+        return value;
```
Comment (tombentley) [inline user-operator/src/main/java/io/strimzi/operator/user/UserOperatorConfig.java]: This version will avoid needing a whole `Map` of all the System properties when you only need one, and also supports a default. I don't think you should rely on the CO setting a value for the env var. 

```java
private static int getIntProperty(String name, int default) {
    String env = System.getEnv(name);
    if (env != null) {
        return Integer.parseInt(env)
    } else {
        return default;
    }
}
```

File: user-operator/src/main/java/io/strimzi/operator/user/UserOperatorConfig.java [hunk 17]
```diff
@@ -124,22 +125,20 @@ public static UserOperatorConfig fromMap(Map<String, String> map) {
     }
 
     public static int getClientsCaValidityDays() {
-        return getIntProperty(UserOperatorConfig.STRIMZI_CLIENTS_CA_VALIDITY);
+        return getIntProperty(UserOperatorConfig.STRIMZI_CLIENTS_CA_VALIDITY, AbstractModel.DEFAULT_CERTS_VALIDITY_DAYS);
     }
 
     public static int getClientsCaRenewalDays() {
-        return getIntProperty(UserOperatorConfig.STRIMZI_CLIENTS_CA_RENEWAL);
+        return getIntProperty(UserOperatorConfig.STRIMZI_CLIENTS_CA_RENEWAL, AbstractModel.DEFAULT_CERTS_RENEWAL_DAYS);
```
Comment (tombentley) [inline user-operator/src/main/java/io/strimzi/operator/user/UserOperatorConfig.java]: Hmm. Having to add the CO as a dependency to access these constants is less than ideal. I suppose the argument could be made that these defaults are part of the CRD API, and so they could be exposed on `KafkaResources`, that would allows you to avoid needing the dependency.

Comment (stanlyDoge) [reply user-operator/src/main/java/io/strimzi/operator/user/UserOperatorConfig.java]: So I should move them to `io/strimzi/api/kafka/model/CertificateAuthority.java`?

File: systemtest/src/test/java/io/strimzi/systemtest/KafkaST.java [hunk 18]
```diff
@@ -944,6 +945,24 @@ void testMirrorMakerTlsScramSha() {
         checkRecordsForConsumer(messagesCount, jobReadMessagesForTarget);
     }
 
+    @Test
+    void testEntityUserOperatorValidityAndRenewal() {
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/KafkaST.java]: Why is this test in `KafkaST.java`? It could be in `KafkaClusterTest` where it will be much quicker to run and fail much earlier.

Comment (stanlyDoge) [reply systemtest/src/test/java/io/strimzi/systemtest/KafkaST.java]: Can we test deployments here? AFAICS it tests only kafka part of cluster.

Comment (tombentley) [reply systemtest/src/test/java/io/strimzi/systemtest/KafkaST.java]: True, it's not `KafkaCluster`, and it looks like there's not a test class which lets you mane assertions about the `Deployment`. But the `EntityUserOperatorTest` would let you asset about the container, and that's enough for env vars.
